### PR TITLE
Feat/shipping state machine

### PR DESCRIPTION
**Pull request by thiagomurakami**
#### What problem is this solving?

Simplify and centralize logic by using xstate to improve code maintenance and quality.

#### How should this be manually tested?

[Workspace](https://shippingstatemachine--checkoutio.myvtex.com/)

#### Checklist/Reminders

- [ ] Updated `README.md`.
- [ ] Updated `CHANGELOG.md`.
- [ ] Linked this PR to a Clubhouse story (if applicable).
- [ ] Updated/created tests (important for bug fixes).
- [ ] Deleted the workspace after merging this PR (if applicable).

#### Type of changes

✔️ | Type of Change
---|---
_ | Bug fix 
_ | New feature 
_ | Breaking change 
✔️ | Technical improvements 

#### Notes

I didn't add a changelog entry because I'm not merging into `master`.

I also recommend installing [XState Devtools](https://github.com/amitnovick/xstate-devtools) (unfortunately, only available for Chrome atm).

### CODE REVIEW

File: react/ShippingForm.tsx [hunk 1]
```diff
@@ -26,87 +19,81 @@ const ShippingForm: React.FC = () => {
...
     }
-    case ShippingStage.EDIT_RECEIVER_INFO: {
+    case matches('editReceiverInfo'): {
       return (
-        <ReceiverInfoForm
-          onReceiverInfoSave={() =>
-            setCurrentStage(ShippingStage.SELECT_DELIVERY_OPTION)
-          }
-        />
+        state.context.selectedAddress && (
```
Comment (lucasecdb) [inline react/ShippingForm.tsx:60]: is it possible for the user to be editing the receiver info without a selected address?

File: react/shippingStateMachine/typings.d.ts [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+type MaybeState<T, U> = U extends true ? Record<'states', T> : T | keyof T
+
+type DeepPartial<T> = {
+  [P in keyof T]?: T[P] extends Array<infer U>
+    ? Array<DeepPartial<U>>
+    : T[P] extends ReadonlyArray<infer U>
+    ? ReadonlyArray<DeepPartial<U>>
+    : DeepPartial<T[P]>
+}
```
Comment (lucasecdb) [inline react/shippingStateMachine/typings.d.ts:11]: :nazare:

File: react/shippingStateMachine/shippingStateMachine.ts [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+      hasIncompleteSelectedAddress: ({ selectedAddress }) =>
+        !!selectedAddress && selectedAddress.addressType == null,
+      buyerIsReceiver: (_, event) => {
+        if (event.type === 'done.invoke.tryToUpdateCompleteAddress') {
+          return event.data.buyerIsReceiver
+        }
+        return false
+      },
+    },
+    services: {
```
Comment (lucasecdb) [inline react/shippingStateMachine/shippingStateMachine.ts]: will we remove these mocked services?

**Review by lucasecdb — APPROVED** [review]
looks good! just a few comments

**Comment by lucasecdb** [comment]
@thiagomurakami when the time comes, could you merge this with the "rebase" option?

File: react/ShippingForm.tsx (hunk 1 above)
Comment (thiagomurakami) [reply react/ShippingForm.tsx:60]: It shouldn't be possible to

**Pull request merged**